seaworthy/tests-core/test_logs.py

```diff
@@ -117,47 +120,90 @@ class FakeLogsContainer:
     """
     A container object stub that emits canned logs.
 
-    Logs can either be streamed or tailed. After a log entry has been streamed,
-    it is stored and we don't wait for it next time. Only logs that have
-    already been streamed are tailed.
+    Logs can either be streamed through a socket using ``.attach_socket()`` or
+    tailed by using ``.logs()``. After a log entry has been streamed, it is
+    stored and we don't wait for it next time. Only logs that have already been
+    streamed are tailed.
     """
 
-    def __init__(self, log_entries, expected_kw=None):
+    def __init__(self, log_entries, expected_params=None):
         self.log_entries = log_entries
         self._seen_logs = []
-        self._expected_kw = {} if expected_kw is None else expected_kw
+        self._expected_params = {
+            'stdout': 1,
+            'stderr': 1,
+            'stream': 1,
+            'logs': 1,
+        }
+        if expected_params is not None:
+            self._expected_params.update(expected_params)
+        self._feeder = None
+
+    def cleanup(self):
+        if self._feeder is not None:
+            self._feeder.cancel()
+            self._feeder = None
 
     def logs(self, stream=False, **kw):
-        if stream:
-            # We're streaming logs, so return our iterator.
-            assert kw == self._expected_kw
-            return self.iter_logs()
-        # We're not streaming logs, make sure we're tailing them.
+        assert stream is False
         tail = kw.get('tail', 'all')
         if tail == 'all':
             tail = len(self._seen_logs)
         assert tail > 0
         return b''.join(self._seen_logs[-tail:])
 
-    def iter_logs(self):
-        for line in self._seen_logs:
-            yield line
-        for wait, line in self.log_entries[len(self._seen_logs):]:
-            sleep(wait)
-            self._seen_logs.append(line)
-            yield line
+    def attach_socket(self, params):
+        assert self._feeder is None
+        assert params == self._expected_params
+        server, client = socket.socketpair()
+        self._feeder = LogFeeder(self, server)
+        self._feeder.start()
+        return socket.SocketIO(client, 'rb')
+
+
+class LogFeeder(threading.Thread):
+    def __init__(self, container, sock):
+        super().__init__()
+        self.con = container
+        self.sock = sock
+        self.finished = threading.Event()
+
+    def cancel(self):
+        self.finished.set()
+        self.sock.close()
+
+    def send_line(self, line):
+        data = b'\x00\x00\x00\x00' + struct.pack('>L', len(line)) + line
+        self.sock.send(data)
+
+    def run(self):
+        for line in self.con._seen_logs:
+            self.send_line(line)
+        for wait, line in self.con.log_entries[len(self.con._seen_logs):]:
+            self.finished.wait(wait)
+            if self.finished.is_set():
+                break
+            self.con._seen_logs.append(line)
```

Comment:
It's only modified by one of them, and that's append-only. Also, I'm pretty sure appending to a list is sufficiently atomic that the GIL protects us even if the caller goes out of its way to have both going at once.